modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/PredicateExpressionVisitorImpl.java

```diff
@@ -321,37 +482,38 @@ else if (Objects.equals(BinaryExpression.Operation.OR, operation)) {
 			return Optional.of(predicate);
 		}
 
-		Column<?, Object> column = _getColumn(left);
-
-		Object value = _getValue(left, right);
+		if (_objectRelationship != null) {
+			try {
+				return Optional.ofNullable(
+					_getPredicateForRelationships(operation, left, right));
+			}
+			catch (Exception exception) {
+				_logExceptionInDebugMode(exception);
 
-		if (Objects.equals(BinaryExpression.Operation.EQ, operation)) {
-			predicate = column.eq(value);
-		}
-		else if (Objects.equals(BinaryExpression.Operation.GE, operation)) {
-			predicate = column.gte(value);
-		}
-		else if (Objects.equals(BinaryExpression.Operation.GT, operation)) {
-			predicate = column.gt(value);
-		}
-		else if (Objects.equals(BinaryExpression.Operation.LE, operation)) {
-			predicate = column.lte(value);
-		}
-		else if (Objects.equals(BinaryExpression.Operation.LT, operation)) {
-			predicate = column.lt(value);
-		}
-		else if (Objects.equals(BinaryExpression.Operation.NE, operation)) {
-			predicate = column.neq(value);
+				return Optional.empty();
+			}
 		}
-		else {
-			return Optional.empty();
+
+		return Optional.ofNullable(
+			_getExpressionPredicate(
+				_getColumn(left, _objectDefinitionId), operation,
+				_getValue(_objectDefinitionId, left, right)));
+	}
+
+	private long _getRelatedObjectDefinitionId(
+		long objectDefinitionId, ObjectRelationship objectRelationship) {
+
+		if (objectRelationship.getObjectDefinitionId1() != objectDefinitionId) {
+			return objectRelationship.getObjectDefinitionId1();
 		}
 
-		return Optional.of(predicate);
+		return objectRelationship.getObjectDefinitionId2();
 	}
 
-	private Object _getValue(Object left, Object right) {
-		EntityField entityField = _getEntityField(left);
+	private Object _getValue(
+		long objectDefinitionId, Object left, Object right) {
```

Comment:
Maybe we should sort the parameters alphabetically

Reply:
In this case, it is intentional. I saw other methods that also receive `left` and `right` and they are at the end. For example:
```
private Optional<Predicate> _getPredicateOptional(
		BinaryExpression.Operation operation, Object left, Object right) {

		Predicate predicate = null;
...
```
So, I decided to leave it this way because I think makes more sense 🤷

Reply:
Good luck with it